Add detection params (userAddress, chainId) and remove duplicate source of truth

wachunei opened the pull request:
**Description**
- REMOVED
  - [BREAKING] Both `collectibles` and `collectibleContracts` are removed from `CollectiblesController` state. 
- FIXED:

	- [BREAKING] Change auto detection from `boolean` to an object containing `address` and `chainId` that was used to autodetect.

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

**Issue**

Related https://github.com/MetaMask/metamask-mobile/issues/3335

gantunesr commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -353,10 +365,7 @@ export class CollectibleDetectionController extends BaseController<
           /* istanbul ignore else */
-          if (
-            !ignored &&
-            requestedSelectedAddress === this.config.selectedAddress
```
Interesting, so this check was only added on the `CollectibleDetectionController` and not in the `CollectiblesController`

gantunesr commented on src/assets/CollectiblesController.ts:
```diff
@@ -118,6 +118,12 @@ export interface CollectibleMetadata {
   collectionImage?: string;
 }
 
+interface DetectionParams {
+  autodetected: boolean;
+  address: string;
+  chainId: string;
+}
+
```
nit: Should this `interface` be at the `CollectibleDetectionController` file?

wachunei replied:
Good question, I'd say since it is used in this controller it should be here.
For example someone wants to use a different method to autodetect: let's say a button which performs a simple API call and with the response data they call `addCollectible`, they need to use that interface.

gantunesr commented on src/assets/CollectiblesController.ts:
```diff
@@ -466,19 +470,32 @@ export class CollectiblesController extends BaseController<
+      const { allCollectibles } = this.state;
+      let chainId, selectedAddress;
+
+      if (detection?.autodetected) {
```
Will there ever be a case where `detection.autodetected = false`?

wachunei replied:
We don't know. This leaves the options open for a developer to just flip a boolean instead of passing or not a whole object.

wachunei replied:
Maybe make it optional and `true` by default? `if (detection && (detection?.autodetected ?? true) ) {`

gantunesr replied:
I don't think its necessary, I do think it could be confusing since the object is already called `detection` which I would assume is being used in the autodetection context and therefore it would make the attribute `autodetected` redundant.

wachunei replied:
I'll remove it and use the presence as condition for branching

gantunesr reviewed (commented):
Left some small comments

adonesky1 commented on src/assets/CollectibleDetectionController.test.ts:
```diff
@@ -289,10 +298,14 @@ describe('CollectibleDetectionController', () => {
-      selectedAddress: '0x1',
+      selectedAddress,
     });
+    collectiblesController.configure({ selectedAddress });
```
I know these are unit tests so we don't need to test the integration points with other Controllers but I generally like to use the upstream method where the piece of state we're subscribed to actually changes (in this case it would be preferences.setSelectedAddress(selectedAddress)). I think it makes our network of assumptions safer but I'm open to being convinced otherwise! Thoughts @Gudahtt @wachunei @gantunesr

Gudahtt replied:
If we wanted to test this robustly we'd do both. I'd rather us migrate to using the new base controller than waste our time doing that though, because we don't really want `configure` to be public API like this.

Following your suggestion would be better at least in that it'd be more compatible with BaseControllerV2 when we do migrate.

adonesky1 commented on src/assets/CollectibleDetectionController.test.ts:
```diff
@@ -328,50 +346,79 @@ describe('CollectibleDetectionController', () => {
     expect(collectiblesController.state.ignoredCollectibles).toHaveLength(0);
     collectiblesController.removeAndIgnoreCollectible(
-      '0x1d963688fe2209a98db35c67a041524822cf04ff',
+      '0x1d963688FE2209A98dB35C67A041524822Cf04ff',
```
just curious why was the checksum change here necessary?

wachunei replied:
Not really, I did it without noticing 😅

adonesky1 commented on src/assets/CollectibleDetectionController.test.ts:
```diff
@@ -328,50 +346,79 @@ describe('CollectibleDetectionController', () => {
     await new Promise((res) => setTimeout(() => res(true), 1000));
     expect(collectibleDetection.config.selectedAddress).toStrictEqual('0x12');
-    expect(collectiblesController.state.collectibles).toStrictEqual([]);
+
+    expect(
+      collectiblesController.state.allCollectibles[
+        collectibleDetection.config.selectedAddress
+      ]?.[chainId] || [],
+    ).toStrictEqual([]);
```
[nit] ```suggestion
const { chainId, selectedAddress } = collectibleDetection.config;

    expect(
      collectiblesController.state.allCollectibles[selectedAddress]?.[chainId],
    ).toStrictEqual([]);
```
why was `|| []` in there? seems like it undermines the test?

wachunei replied:
Here's the reason: previous tests were against these 'calculated properties', these were always an array, that's why the tests that are expecting emptiness are using `toStrictEqual([])`.

Emptiness in the state now is represented by the absence of the key (address or chainId), for example if you switch to `selectedAddress` `'0x2'` and `chainId` `'56'`, the data structure will not have that nested object, ie: `collectiblesController.state.allCollectibles['0x2']['56']` will throw because `collectiblesController.state.allCollectibles['0x2']` will be undefined, since there are no collectibles.

The `|| []` is used as fallback to match the previous tests, this is also used in the state selectors on mobile, because components props are expecting an array. (https://github.com/MetaMask/metamask-mobile/pull/3458/files#diff-065d8fc10fb71f841b46676cb75d0242d8ec471c6a912ce1142e4ba121fee5c1R12-R30)

We could definitely change the expectation to be `undefined` or `[]`  in according on each test case.

adonesky1 replied:
Ok yeah I think it makes more sense to change the expectation to `undefined` because that is actually what we expect right? Rather than changing it to an empty array in the test itself if `undefined`.

adonesky1 replied:
And in many of these cases you've included the `|| []` we don't even expect them to be `undefined`

wachunei replied:
Right, there are some tests that add, expect, remove, expect, in that case the array existed and now it should be `[]` instead of `undefined`. But I'll update them that way.

adonesky1 commented on src/assets/CollectibleDetectionController.test.ts:
```diff
@@ -328,50 +346,79 @@ describe('CollectibleDetectionController', () => {
+    const { chainId } = collectiblesController.config;
     collectibleDetection.detectCollectibles();
-    expect(collectiblesController.state.collectibles).toStrictEqual([]);
+    expect(
```
same question as above?

adonesky1 commented on src/assets/CollectibleDetectionController.test.ts:
```diff
@@ -429,19 +476,26 @@ describe('CollectibleDetectionController', () => {
+    expect(
+      collectiblesController.state.allCollectibles[selectedAddress]?.[
+        chainId
+      ] || [],
```
what am I missing about these instances of`|| []`? I don't see why they are needed?

adonesky1 commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -131,6 +132,7 @@ export interface ApiCollectibleCollection {
 export interface CollectibleDetectionConfig extends BaseConfig {
   interval: number;
   networkType: NetworkType;
+  chainId: `0x${string}` | `${number}` | number;
```
Whoa! I didn't know you could do this! So cool

adonesky1 commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -309,15 +314,22 @@ export class CollectibleDetectionController extends BaseController<
 
     await safelyExecute(async () => {
-      const apiCollectibles = await this.getOwnerCollectibles();
+      const apiCollectibles = await this.getOwnerCollectibles(selectedAddress);
```
I'm not understanding what we achieve by getting `selectedAddress` in `detectCollectibles` and passing it in to `getOwnerCollectibles` rather than just pulling `selectedAddress` off config in `getOwnerCollectibles`?

wachunei replied:
We just want to avoid grabbing it in different places to make sure it does not change along the execution of the procedure

adonesky1 commented on src/assets/CollectiblesController.test.ts:
```diff
@@ -427,15 +461,25 @@ describe('CollectiblesController', () => {
 
-      expect(collectiblesController.state.collectibles[0]).toStrictEqual({
+      expect(
+        collectiblesController.state.allCollectibles[selectedAddress]?.[
+          NetworksChainId[secondNetworkType]
+        ] || [],
+      ).toHaveLength(0);
```
```suggestion
        collectiblesController.state.allCollectibles[selectedAddress]?.[
          NetworksChainId[secondNetworkType]
        ]
      ).toBe(undefined);
```

adonesky1 commented on src/assets/CollectiblesController.test.ts:
```diff
@@ -598,18 +685,29 @@ describe('CollectiblesController', () => {
+        collectiblesController.state.allCollectibleContracts[selectedAddress][
+          chainId
+        ],
+      ).toHaveLength(0);
```
Should we remove the key rather than just remove the item from the array when there are no more left for the address/chainId combo?

Gudahtt replied:
I agree that this would be nicer, so that the behaviour after adding and removing a collectible matches the initial behaviour. Leaves us with less cases to test, and it's simpler for anyone using this state. I don't want to block on it though.

adonesky1 commented on src/assets/CollectiblesController.ts:
```diff
@@ -466,19 +469,32 @@ export class CollectiblesController extends BaseController<
    * @param address - Hex address of the collectible contract.
    * @param tokenId - The collectible identifier.
    * @param collectibleMetadata - Collectible optional information (name, image and description).
+   * @param detection - Wether the collectible is manually added or auto-detected for address and chainId.
```
```suggestion
   * @param detection - An object containing the users currently selected address and the chainId used to ensure detected collectibles are added to the correct account.
```

adonesky1 commented on src/assets/CollectiblesController.ts:
```diff
@@ -118,6 +118,11 @@ export interface CollectibleMetadata {
 }
 
+interface DetectionParams {
+  address: string;
```
[nit] - we should distinguish this from the contractAddress for clarity
```suggestion
  userAddress: string;
```

wachunei replied:
Renamed in a80f003

adonesky1 commented on src/assets/CollectiblesController.ts:
```diff
@@ -535,18 +550,31 @@ export class CollectiblesController extends BaseController<
    *
    * @param address - Hex address of the collectible contract.
-   * @param detection - Whether the collectible is manually added or auto-detected.
+   * @param detection - Whether the collectible is manually added or auto-detected for address and chainId.
```
```suggestion
   * @param detection - An object containing the users currently selected address and the chainId used to ensure detected collectibles are added to the correct account.
```

adonesky1 commented on src/assets/CollectiblesController.ts:
```diff
@@ -906,14 +919,14 @@ export class CollectiblesController extends BaseController<
    * @param tokenId - The collectible identifier.
    * @param collectibleMetadata - Collectible optional metadata.
-   * @param detection - Whether the collectible is manually added or autodetected.
+   * @param detection - Whether the collectible is manually added or auto-detected for address and chainId.
```
```suggestion
   * @param detection - An object containing the users currently selected address and the chainId used to ensure detected collectibles are added to the correct account.
```

adonesky1 reviewed (changes requested):
I think in general I don't like the pattern of replacing `undefined` with an empty array because it maintains the default value of the `collectibles` and `collectibleContracts` subset state values. I think it ends up causing more confusion. Open to be convinced otherwise though!

wachunei commented:
I updated the `expect` arguments to be assertive (removed optional chaining when we expect a value, removed default `[]`) and the expectations to match this change.

About this:

> [Should we remove the key rather than just remove the item from the array when there are no more left for the address/chainId combo?](https://github.com/MetaMask/controllers/pull/636#discussion_r757689298)

I think it makes sense to prune empty nodes, is it something we do usually? I'd defer decision to you guys that are more familiar with controllers conventions.

adonesky1 reviewed (approved):
LGTM

Gudahtt commented:
So these "detection params", what purpose do these serve? Is this so that the chainID/address at time of detection is preserved, so that the NFT is assigned correctly if the user switches accounts/networks while the detection was still being processed? I didn't see the problem this was meant to solve explained anywhere.

wachunei commented:
Yes, exactly that

Gudahtt commented on src/assets/CollectiblesController.ts:
```diff
@@ -466,19 +469,32 @@ export class CollectiblesController extends BaseController<
    * @param address - Hex address of the collectible contract.
    * @param tokenId - The collectible identifier.
    * @param collectibleMetadata - Collectible optional information (name, image and description).
+   * @param detection - An object containing the users currently selected address and the chainId used to ensure detected collectibles are added to the correct account.
```
This description seems wrong. `this.config.chainId` and `this.config.selectedAddress` are the users currently selected network and address. If that's what this was intended to be, we could just use the config.

Would this be more accurate?
```suggestion
   * @param detection - The chain ID and address of the currently selected network and account at the moment the collectible was detected.
```

Gudahtt replied:
I guess this same description could be used for `addCollectibleContract` and `addCollectible` as well

wachunei replied:
Changed in d7f56d0c93718842f59fb23002cafbf1d4e70162

Gudahtt commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -381,12 +390,10 @@ export class CollectibleDetectionController extends BaseController<
-            );
+            await this.addCollectible(address, token_id, collectibleMetadata, {
+              userAddress: selectedAddress,
+              chainId: chainId as string,
```
Nit: We should not need to cast here. If it's possible for this to be something other than a string, we need to deal with that here. If it's not, then the type is wrong and we should fix that instead.

Gudahtt replied:
The chain ID normalization above is responsible for this. It is guaranteed to be a string here but the normalization is done in a way TypeScript's inference is having trouble with. I've just prototyped a solution that I can prepare as a follow up PR later.

Gudahtt replied:
Basically it's to move the normalization to a utility function, and adjust the conditions to help TypeScript's inference along a bit better, and add a "throw" case for invalid chain IDs.

wachunei replied:
I see, this type and normalization is also present in the `GasFeeController`, so we can use that util there.

Gudahtt commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -249,7 +251,10 @@ export class CollectibleDetectionController extends BaseController<
-      this.configure({ networkType: provider.type });
+      this.configure({
+        networkType: provider.type,
+        chainId: provider.chainId as CollectibleDetectionConfig['chainId'],
```
Nit: Ideally we'd parse this here to guarantee that the chain ID is the expected type, rather than cast. Or propagate this new chain ID type into the network controller as well. I guess we can do that later.

Gudahtt commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -309,15 +314,22 @@ export class CollectibleDetectionController extends BaseController<
+
+    let { chainId } = this.config;
+    if (typeof chainId === 'string' && isHexString(chainId)) {
+      chainId = `${parseInt(chainId, 16)}` as `${number}`;
+    } else if (typeof chainId === 'number') {
+      chainId = `${chainId}` as `${number}`;
```
Using `as const` here is enough to get TypeScript to accept these values as template literal types. Const assertions are safer than casting directly to the expected type, because it still lets `tsc` type check this for us.
```suggestion
      chainId = `${parseInt(chainId, 16)}` as const;
    } else if (typeof chainId === 'number') {
      chainId = `${chainId}` as const;
```

wachunei replied:
This is what I needed and I did not know it, thanks! I will commit it directly from github.

Gudahtt commented on src/assets/CollectibleDetectionController.ts:
```diff
@@ -228,6 +229,7 @@ export class CollectibleDetectionController extends BaseController<
     this.defaultConfig = {
       interval: DEFAULT_INTERVAL,
       networkType: MAINNET,
+      chainId: '1',
```
Nit: I don't object to doing this for now, since the default `networkType` is `MAINNET` already, so this is a pre-existing issue. But we shouldn't be setting a default network at all, we can get the true current network from the network controller by passing in a `getNetworkState` function to the constructor.

This could lead to bugs where something runs initially before the first network state update, and uses the wrong network.

Gudahtt commented:
Overall looks good! I'm very glad to see that duplicate state removed. It would have been better to do this in two separate PRs, but what's done is done. I can take another look when conflicts are resolved.

wachunei commented:
@Gudahtt Conflicts are resolved 👌

Gudahtt reviewed (approved):
LGTM!
